src/main/utility/proc_maps.rs

```diff
@@ -0,0 +1,470 @@
+use regex::Regex;
+use std::error::Error;
+use std::fmt::Display;
+use std::str::FromStr;
+
+/// Whether a region of memory is shared.
+#[derive(PartialEq, Eq, Debug, Copy, Clone)]
+pub enum Sharing {
+    Private,
+    Shared,
+}
+
+impl FromStr for Sharing {
+    type Err = Box<dyn Error>;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == "p" {
+            Ok(Sharing::Private)
+        } else if s == "s" {
+            Ok(Sharing::Shared)
+        } else {
+            Err(format!("Bad sharing specifier {}", s))?
+        }
+    }
+}
```

Comment:
I've been trying to understand how this works but I'm confused. Supposedly (using the `print_type_of` function in https://stackoverflow.com/a/58119924) the `Err(format!("Bad sharing specifier {}", s))` is of type `Result<Result<Sharing, Box<dyn Error>>, String>`, but I don't see how the `?` turns it into a `Result<Sharing, Box<dyn Error>>`. And since `String` does not implement `Error`, I don't see how the `String` can be returned in the result. It works, but I have no idea why :)

Reply:
There's an implementation of `From<String> for Box<dyn Error>`: https://doc.rust-lang.org/std/string/struct.String.html#impl-From%3CString%3E-6

The `?` operator uses that to do the conversion :)

Reply:
Ah okay that makes things much clearer, thanks :) And it's great how the Rust documentation links to the source code, so you can see that the conversion creates a `StringError` type.